### PR TITLE
[Sessions][1/3] Support multiple sessions

### PR DESCRIPTION
**Pull request by yuna878**
## Overview

Support multiple sessions for a user.

## Changes Made

Abstracted session related data from `User` model and put into the new `Session` model.
Refer to README for detailed description.

## Test Coverage

Pytest for general functionality & manual testing for push notifications

## Next Steps (delete if not applicable)

- Handle invalid device token errors
- Script to periodically clean up outdated sessions

### CODE REVIEW

File: src/app/coursegrab/controllers/update_notification_controller.py [hunk 1]
```diff
@@ -13,5 +13,7 @@ def get_methods(self):
     def content(self, **kwargs):
         data = request.get_json()
         user = kwargs.get("user")
+
+        # Cast "NONE" string to null value
```
Comment (aliciaxw) [inline src/app/coursegrab/controllers/update_notification_controller.py:17]: might be worth specifying that `NONE` is a constant coming from `constants.py`, the name could be a bit confusing for python newcomers (`None` vs `NONE`)

File: src/app/coursegrab/models/user.py [hunk 2]
```diff
@@ -17,44 +15,22 @@ class User(db.Model):
...
-
-    session_token = db.Column(db.String, nullable=False, unique=True)
-    session_expiration = db.Column(db.DateTime, nullable=False)
-    update_token = db.Column(db.String, nullable=False, unique=True)
-
     notification = db.Column(db.String, nullable=True)
-    device_token = db.Column(db.String, nullable=True)
 
     sections = db.relationship("Section", secondary=users_to_sections, backref="users")
+    sessions = db.relationship("Session", back_populates="user")
```
Comment (aliciaxw) [inline src/app/coursegrab/models/user.py:21]: do we want to cascade delete sessions if the user is deleted?

File: src/app/coursegrab/utils/push_notifications.py [hunk 3]
```diff
@@ -26,13 +27,15 @@ def notify_users(section):
...
         if user.notification == ANDROID:
-            android_tokens.append(user.device_token)
-        elif user.notification == IPHONE:
-            ios_tokens.append(user.device_token)
-
+            android_tokens.extend(get_user_device_tokens(user.id, ANDROID))
+        elif user.notification == IOS:
+            ios_tokens.extend(get_user_device_tokens(user.id, IOS))
+        # elif user.notification == EMAIL:
+        #     emails.append(user.e)
```
Comment (aliciaxw) [inline src/app/coursegrab/utils/push_notifications.py:38]: consider creating a github issue for implementing email and then tagging these code comments with the issue # so it's easy to find where you started implementing

File: src/tests/test_coursegrab.py [hunk 4]
```diff
@@ -99,25 +106,46 @@ def test_search_course(client, user):
...
+
+    assert new_session.id == session.id
+    assert new_session.session_token == session.session_token
+    assert new_session.device_token == device_token
+
+
+def test_create_session_no_device_token(client, session, user):
+    sessions_dao.create_session(user.id, session.device_type)
+
+    assert len(user.sessions) == 2
```
Comment (aliciaxw) [inline src/tests/test_coursegrab.py:151]: we love this

**Review by aliciaxw — APPROVED** [review]
glorious PR 🔥  minor suggestions

File: src/app/coursegrab/utils/constants.py [hunk 5]
```diff
@@ -9,9 +9,12 @@
 
 # Possible values for notification
 ANDROID = "ANDROID"
-IPHONE = "IPHONE"
+IOS = "IOS"
 EMAIL = "EMAIL"
 NONE = "NONE"
 
+# Possible values for device type
+WEB = "WEB"  # + ANDROID, IOS
```
Comment (alanna-zhou) [inline src/app/coursegrab/utils/constants.py:17]: is the comment for android/ios ppl logged in on web

File: src/tests/test_coursegrab.py [hunk 6]
```diff
@@ -20,77 +21,83 @@ def client():
 
 @pytest.fixture
 def user():
-    _, user = users_dao.create_user("test@email.com", "First", "Last")
+    user = users_dao.create_user("test@email.com", "First", "Last")
```
Comment (alanna-zhou) [inline src/tests/test_coursegrab.py:24]: seems pretty straightforward, would probs be cleaner if we just returned `users_dao.create_user...`

File: src/tests/test_coursegrab.py [hunk 7]
```diff
@@ -20,77 +21,83 @@ def client():
...
     return user
 
 
-def test_hello_world(client, user):
-    res = client_get(client, user, "/api/hello/")
+@pytest.fixture
+def session(user):
+    session = sessions_dao.create_session(user.id, IOS)
+    return session
```
Comment (alanna-zhou) [inline src/tests/test_coursegrab.py:31]: same here

**Review by alanna-zhou — APPROVED** [review]
damnnn a thiccccc pr 💯 you also included tests :O impressive!!!

**Review by manishsaha — APPROVED** [review]
this looks great! agree with the points alicia and alanna have made.

File: src/app/coursegrab/utils/constants.py (hunk 5 above)
Comment (yuna878) [reply src/app/coursegrab/utils/constants.py:17]: It means that possible values for device type is `WEB || ANDROID || IOS`

**Pull request merged**